### DISCUSSION TITLE
Calculating the ascendant and midheaven

### DISCUSSION THREAD

**Discussion by KurtPreston** [cosinekitty/astronomy#340] [Q&A] [answered]
I am trying to calculate the ascendant and midheaven astrological points. My understanding is that the ascendant is calculated by determining the ecliptic longitude of due east on the horizon at a specific place and time. The midheaven is the ecliptic longitude of the zenith.

I have tried implementing this using astronomy-engine, but my implementation seems to produce incorrect results. I am unsure whether it is a mistake in my understanding of the astrology, or if it's a bug in the code, but any help would be appreciated.

```typescript
export function calculateAscendant(params: {
  latitude: number;
  longitude: number;
  date: Date;
}) {
  const {latitude, longitude, date} = params;
  const observer = new Astronomy.Observer(latitude, longitude, 0);
  const horizVector: Astronomy.Vector = new Astronomy.Vector(
    0, // north-south
    1, // west-east
    0, // up-down
    new Astronomy.AstroTime(date)
  );
  const rotation = Astronomy.Rotation_HOR_EQJ(date, observer);
  const eqjVector = Astronomy.RotateVector(rotation, horizVector);
  const ecliptic = Astronomy.Ecliptic(eqjVector);
  return ecliptic.elon;
}
```

To get the midheaven, I instead initialize the `horizVector` with this:

```typescript
  const horizVector: Astronomy.Vector = new Astronomy.Vector(
    0, // north-south
    0, // west-east
    1, // up-down
    new Astronomy.AstroTime(date)
  );
```

Note, a similar question was asked before [here](https://github.com/cosinekitty/astronomy/discussions/166).

Also, thank you for this library! I love this project.

**Comment by cosinekitty** [comment]
So far your approach looks mostly correct.

I spotted one problem with the vector you intended to point due east. It actually points west: x = north, y = west, z = up. So the first vector should be (0, -1, 0).

Your midheaven calculation looks correct. Did it give you the expected answer?

**Reply by KurtPreston** [reply]
Much appreciate the reply! I have updated the ascendant `horizVector`, but something still seems a bit off.

After trying several more data points, it seems to be producing values in the range of 2° to 20° different from other astrology programs for both ascendant and midheaven. I'm not sure the source of the discrepancy.

For example, see this calculation for Salvador Dalli's ascendant:

```typescript
import * as Astronomy from 'astronomy-engine';

const latitude = 42.267;
const longitude = 2.967;
const date = new Date('1904-05-11T08:45:00Z')
const observer = new Astronomy.Observer(latitude, longitude, 0);
const horizVector: Astronomy.Vector = new Astronomy.Vector(
  0, // north-south
  -1, // west-east
  0, // down-up
  new Astronomy.AstroTime(date)
);
const rotation = Astronomy.Rotation_HOR_EQJ(date, observer);
const eqjVector = Astronomy.RotateVector(rotation, horizVector);
const ecliptic = Astronomy.Ecliptic(eqjVector);
console.log(ecliptic.elon); // Returns 93.187 (3.187°Cn)
```

When I plug that lat/lon/time into other Astrology programs, here is what they return:
- AstroGold: 22° Cn 14'41" (or 112.245°, a 19° difference)
- AstroMatrix: 22° Cn 14'
- CafeAstrology.com: 14° Cn 57' 15"

I don't know how these programs are internally calculating the value, but they do seem closer to agreeing with each other than with the value returned from my function.

**Reply by KurtPreston** [reply]
I have a little clarification on what is happening. There is a formula commonly used for calculating the ascendant described by Wikipedia [here](https://en.wikipedia.org/wiki/Ascendant).

When I use this formula to calculate the ascendant instead of the matrix transformations, the results are extremely close to the values reported by astrology programs. I am assuming these programs are probably using this formula.

I am not expert in astronomical calculations, but I am wondering—which system produces a more accurate result? astronomy-engine seems to take a more thorough approach to validating its data, but I wonder if the cumulative effect of all the individual transformation can cause more lost precision than an optimized formula? Or is the formula provided by Wikipedia a crude approximation of some much more complicated spatial geometry that is more accurately handled by the matrix transformations?

**Reply by gsvjk** [reply]
@KurtPreston hi kurt, using your code i get correct values. i just verified with jhora.

**Reply by cosinekitty** [reply]
@KurtPreston, I'm looking at the formula on the Wikipedia page you mentioned, specifically the one from Meeus 1991:

<img src="https://wikimedia.org/api/rest_v1/media/math/render/svg/68649ded47b9e668ef643fbf313d157ed4037022" />

It may take me a day or two, but I will study this formula so that I can understand it and figure out why it's giving a different answer than the rotation matrix approach. It is possible the Meeus formula answers a different question. It is also possible we are all looking at this wrong somehow...

**Reply by KurtPreston** [reply]
@cosinekitty Thank you for offering to look into it, but after scratching my head over this for a few days, I realize I had the definition of ascendant slightly wrong.

The horizontal vector should not be pointing due east, but rather aimed at the point of the Eastern horizon intersecting with the ecliptic. I'm not totally certain how I should calculate that angle yet, but it would explain the discrepancy.

If I can fix the matrix implementation, I'll do a comparison of values to see how close it is to the angle returned from the Meeus formula.

Any advice on how to calculate the horizontal vector with the correct azimuth would be appreciated. Again, thanks!

**Reply by cosinekitty** [reply]
Yes, this is an interesting problem! The new definition makes sense, but it's not yet obvious how to calculate it. Let me think about it...

**Reply by cosinekitty** [reply]
So far, the best approach I can figure is to implement a search. It would scan around the eastern horizon, with the azimuth as a free variable in the range (0, +180). From this azimuth, produce a horizontal vector: (variable azimuth, altitude = 0, distance = 1). Convert that horizontal vector to an ecliptic vector.

Here you have options. If you convert to a cartesian ecliptic vector (ex, ey, ez), you can search for whatever azimuth makes ez closest to zero.

If you convert to spherical coordinates (elon, elat, distance), you can search for the azimuth that makes elat closest to zero.

Astronomy Engine provides a customizable search function [Search](https://github.com/cosinekitty/astronomy/tree/master/source/js#searchfunc-t1-t2-options--astrotime--null), but unfortunately, it is geared for time as the free variable, not azimuth angles. 

Overall, I think the Meeus formula is the best solution for this particular application because it is a direct solution rather than a search.

**Comment by KurtPreston** [comment, marked as answer]
Makes sense. In case it proves useful to anyone else, here's an implementation of the Meeus formulas for ascendant and midheaven using astronomy-engine to get the ecliptic obliquity and sidereal time:

```typescript
import * as Astronomy from 'astronomy-engine';

function calculateAscendant(params: {
  latitude: number;
  longitude: number;
  date: Date;
}) {
  const {date, latitude} = params;
  const localSiderealRadians = localSiderealTimeRadians(params);
  const eclipticObliquity = degreesToRadians(
    Astronomy.e_tilt(new Astronomy.AstroTime(date)).tobl
  );
  const x =
    Math.sin(localSiderealRadians) * Math.cos(eclipticObliquity) +
    Math.tan(degreesToRadians(latitude)) * Math.sin(eclipticObliquity);
  const y = -1 * Math.cos(localSiderealRadians);
  const celestialLongitudeRadians = Math.atan(y / x);
  let ascendantDegrees = radiansToDegrees(celestialLongitudeRadians);

  // Correcting the quadrant
  if (x < 0) {
    ascendantDegrees += 180;
  } else {
    ascendantDegrees += 360;
  }
  if (ascendantDegrees < 180) {
    ascendantDegrees += 180;
  } else {
    ascendantDegrees -= 180;
  }
  return ascendantDegrees;
}

function calculateMidheaven(params: {
  latitude: number;
  longitude: number;
  date: Date;
}) {
  const {date} = params;
  const localSiderealRadians = localSiderealTimeRadians(params);
  const eclipticObliquity = degreesToRadians(
    Astronomy.e_tilt(new Astronomy.AstroTime(date)).tobl
  );
  const numerator = Math.tan(localSiderealRadians);
  const denominator = Math.cos(degreesToRadians(eclipticObliquity));
  let midheavenDegrees = radiansToDegrees(Math.atan(numerator / denominator));

  // Correcting the quadrant
  if (midheavenDegrees < 0) {
    midheavenDegrees += 360;
  }
  const localSiderealDegrees = radiansToDegrees(localSiderealRadians);
  if (midheavenDegrees > localSiderealDegrees) {
    midheavenDegrees -= 180;
  }
  if (midheavenDegrees < 0) {
    midheavenDegrees += 180;
  }
  if (midheavenDegrees < 180 && localSiderealDegrees >= 180) {
    midheavenDegrees += 180;
  }
  midheavenDegrees = (midheavenDegrees + 360) % 360;
  return midheavenDegrees;
}

export function localSiderealTimeRadians(params: {
  longitude: number;
  date: Date;
}): number {
  const {longitude, date} = params;
  const grenichSiderealTime = Astronomy.SiderealTime(date);
  const localSiderealTime = grenichSiderealTime + degreesToHours(longitude);
  let localSiderealDegrees = hoursToDegrees(localSiderealTime);
  localSiderealDegrees = (localSiderealDegrees + 360) % 360;
  const localSiderealRadians = degreesToRadians(localSiderealDegrees);
  return localSiderealRadians;
}

function hoursToDegrees(hours: number): number {
  return hours * 15;
}

function degreesToHours(degrees: number): number {
  return degrees / 15;
}

function degreesToRadians(degrees: number): number {
  return degrees / 180 * Math.PI
}

function radiansToDegrees(radians: number): number {
  return radians * 180 / Math.PI;
}
```

**Reply by ebraminio** [reply]
Thanks to your code @KurtPreston but as e_tilt wasn't exposed in the Kotlin port my code turned it like this,

```Kotlin
// Extracted from not exposed calculations of astronomy library
private fun eclipticObliquity(time: Time): Double {
    val t = time.tt / 36525.0
    val asec =
        ((((-0.0000000434 * t
                - 0.000000576) * t
                + 0.00200340) * t
                - 0.0001831) * t
                - 46.836769) * t + 84381.406
    val mobl = asec / 3600
    val tobl = mobl + (time.nutationEps() / 3600)
    return Math.toRadians(tobl)
}

fun calculateAscendant(latitude: Double, longitude: Double, time: Time): Double {
    val localSiderealRadians = localSiderealTimeRadians(longitude, time)
    val eclipticObliquity = eclipticObliquity(time)
    val x = sin(localSiderealRadians) * cos(eclipticObliquity) +
            tan(Math.toRadians(latitude)) * sin(eclipticObliquity)
    val y = -1 * cos(localSiderealRadians)
    val celestialLongitudeRadians = atan(y / x)
    var ascendantDegrees = Math.toDegrees(celestialLongitudeRadians)

    // Correcting the quadrant
    ascendantDegrees += if (x < 0) 180.0 else 360.0
    if (ascendantDegrees < 180.0) ascendantDegrees += 180.0 else ascendantDegrees -= 180.0
    return ascendantDegrees
}

fun calculateMidheaven(longitude: Double, time: Time): Double {
    val localSiderealRadians = localSiderealTimeRadians(longitude, time);
    val eclipticObliquity = eclipticObliquity(time)
    val numerator = tan(localSiderealRadians)
    val denominator = cos(eclipticObliquity)
    var midheavenDegrees = Math.toDegrees(atan(numerator / denominator))

    // Correcting the quadrant
    if (midheavenDegrees < 0) midheavenDegrees += 360
    val localSiderealDegrees = Math.toDegrees(localSiderealRadians)
    if (midheavenDegrees > localSiderealDegrees) midheavenDegrees -= 180
    if (midheavenDegrees < 0) midheavenDegrees += 180
    if (midheavenDegrees < 180 && localSiderealDegrees >= 180) midheavenDegrees += 180
    midheavenDegrees = (midheavenDegrees + 360) % 360
    return midheavenDegrees
}

private fun hoursToDegrees(hours: Double) = hours * 15
private fun degreesToHours(degrees: Double) = degrees / 15

private fun localSiderealTimeRadians(longitude: Double, time: Time): Double {
    val greenwichSiderealTime = siderealTime(time)
    val localSiderealTime = greenwichSiderealTime + degreesToHours(longitude)
    var localSiderealDegrees = hoursToDegrees(localSiderealTime)
    localSiderealDegrees = (localSiderealDegrees + 360) % 360
    val localSiderealRadians = Math.toRadians(localSiderealDegrees)
    return localSiderealRadians
}
```

Maybe exposing earthTilt in the Kotlin port can be nice.

Btw, thanks @cosinekitty for providing this library and maintaining this community. I haven't forgotten your help on providing the Kotlin port which over time has helped the free and open source app I'm maintaining to become unique by it's own. Thanks again

**Reply by ebraminio** [reply]
Turned out there is a small issue in midheaven calculation and that is `degreesToRadians` is being applied to `eclipticObliquity` twice, I've fixed that in my calculation.

**Comment by gsvjk** [comment]
@KurtPreston good job kurt. I ended up implementing the same - asc, midheaven and lunar nodes. I was able to verify asc and lunar nodes but not mid heaven. Any source where i can verify the same ?

**Reply by gsvjk** [reply]
@KurtPreston verified the midheaven calculation [here](https://horoscopes.astro-seek.com/calculate-birth-chart-horoscope-online/?input_natal=1&send_calculation=1&narozeni_den=30&narozeni_mesic=3&narozeni_rok=2024&narozeni_hodina=13&narozeni_minuta=33&narozeni_sekunda=00&narozeni_city=Mumbai%2C+India%2C+Maharashtra&narozeni_mesto_hidden=Mumbai&narozeni_stat_hidden=IN&narozeni_podstat_kratky_hidden=&narozeni_podstat_hidden=Maharashtra&narozeni_input_hidden=&narozeni_podstat2_kratky_hidden=&narozeni_podstat3_kratky_hidden=&narozeni_sirka_stupne=19&narozeni_sirka_minuty=4&narozeni_sirka_smer=0&narozeni_delka_stupne=72&narozeni_delka_minuty=53&narozeni_delka_smer=0&narozeni_timezone_form=auto&narozeni_timezone_dst_form=auto&house_system=placidus&hid_fortune=1&hid_fortune_check=on&hid_vertex=1&hid_vertex_check=on&hid_chiron=1&hid_chiron_check=on&hid_lilith=1&hid_lilith_check=on&hid_uzel=1&hid_uzel_check=on&tolerance=1&aya=&tolerance_paral=1.2#tabs_redraw). midheaven is off by 2 degrees. 

```
        latitude: 18.966666,
        longitude: 72.833333
        2024-03-30T08:03:00.000Z // 2024 30th mar, 13:33 pm +5:30
        
        expected value : 23.54
        returned value: 21.78 // off by 2 degrees
        
```

**Reply by KurtPreston** [reply]
@gsvjk If I change the `eclipticObliquity` to use the J2000 value of `23.4392911` instead of retrieving it from astronomy-engine using `Astronomy.e_tilt`, then the formula outputs `23.54`.

My guess is that astro-seek is using a hard-coded value for the obliquity instead of looking up it's value at the specified date.

**Reply by ebraminio** [reply]
I think midheaven calculation mismatch was due to the need of a small fix here, 

```
  const eclipticObliquity = degreesToRadians(
    Astronomy.e_tilt(new Astronomy.AstroTime(date)).tobl
  );
  const numerator = Math.tan(localSiderealRadians);
  const denominator = Math.cos(degreesToRadians(eclipticObliquity));
```

That turns eclipticObliquity to radians twice.

Thanks again @KurtPreston for sharing your code as it was a solid basis for my implementation in https://github.com/cosinekitty/astronomy/discussions/391
